src/SqlProviderBase.ts

```diff
@@ -889,40 +895,79 @@ class SqlStore implements NoSqlProvider.DbStore {
+                    if (serializedKeys.length > 0) {
+                        if (!dataToInsertByIndex[indexIndex]) {
+                            dataToInsertByIndex[indexIndex] = [];
                         }
-                        return undefined;
-                    }));
+                        const dataToInsert = dataToInsertByIndex[indexIndex];
+                        _.each(serializedKeys, val => {
+                            dataToInsert.push(val);
+                            dataToInsert.push(key);
+                            if (index.includeDataInIndex) {
+                                dataToInsert.push(datas[itemIndex]);
+                            }
+                        });
+                    }
+
+                    // Capture delete keys
+                    if (!keysToDeleteByIndex[indexIndex]) {
+                        keysToDeleteByIndex[indexIndex] = [];
+                    }
+
+                    keysToDeleteByIndex[indexIndex].push(key);
                 });
             });
+
+            const deleteQueries: SyncTasks.Promise<void>[] = [];
+
+            _.each(keysToDeleteByIndex, (keysToDelete, indedIndex) => {
+                // We know indexes are defined if we have data to insert for them
+                // _.each spits dictionary keys out as string, needs to turn into a number
+                const index = this._schema.indexes!!![Number(indedIndex)];
+                const itemPageSize = this._trans.internal_getMaxVariables();
+                for (let i = 0; i < keysToDelete.length; i += itemPageSize) {
+                    const thisPageCount = Math.min(itemPageSize, keysToDelete.length - i);
+                    deleteQueries.push(this._trans.internal_nonQuery('DELETE FROM ' + this._schema.name + '_' + index.name +
+                        ' WHERE nsp_refpk IN (' + generateParamPlaceholder(thisPageCount) + ')', keysToDelete.splice(0, thisPageCount)));
+                }
+            });
+
+            queries.push(...deleteQueries);
+
+            const allDeleteQueries = SyncTasks.all(deleteQueries);
+            
+            _.each(dataToInsertByIndex, (data, indexIndex) => {
+                // We know indexes are defined if we have data to insert for them
+                // _.each spits dictionary keys out as string, needs to turn into a number
+                const index = this._schema.indexes!!![Number(indexIndex)];
+                const insertParamCount = index.includeDataInIndex ? 3 : 2;
+                const itemPageSize = Math.floor(this._trans.internal_getMaxVariables() / insertParamCount);
+                // data contains all the input parameters
+                for (let i = 0; i < (data.length / insertParamCount); i += itemPageSize) {
+                    const thisPageCount = Math.min(itemPageSize, (data.length / insertParamCount)) - i;
+                    const qmarksValues = _.fill(new Array(thisPageCount), generateParamPlaceholder(insertParamCount));
```

Comment:
Suggestion: The same `_.repeat` idea could be used here instead of `_.fill`, `Array`, and `.join`.
```ts
_.repeat('(' + generateParamPlaceholder(insertParamCount) + '),', thisPageCount).slice(0, -1);
```
Perhaps this could be extracted:
```ts
repeatWithSeparator(str: string, n: number, sep: string) {
    return _.repeat(str + sep, n).slice(0, -sep.length);
}
```